src/librbd/image/CreateRequest.cc

```diff
@@ -0,0 +1,795 @@
+    create_rados_ack_callback<klass, &klass::handle_object_map_resize>(this);
+  int r = m_ioctx.aio_operate(m_objmap_name, comp, &op);
+  assert(r == 0);
+  comp->release();
+}
+
+template<typename I>
+Context *CreateRequest<I>::handle_object_map_resize(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "error creating initial object map: " << cpp_strerror(*result) << dendl;
+
+    m_r_saved = *result;
+    remove_header_object();
+    return nullptr;
+  }
+
+  fetch_mirror_mode();
+  return nullptr;
+}
+
+template<typename I>
+void CreateRequest<I>::fetch_mirror_mode() {
+  if ((m_features & RBD_FEATURE_JOURNALING) == 0) {
+    complete(0);
+    return;
+  }
+
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectReadOperation op;
+  cls_client::mirror_mode_get_start(&op);
+
+  using klass = CreateRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_fetch_mirror_mode>(this);
+  m_outbl.clear();
+  int r = m_ioctx.aio_operate(RBD_MIRRORING, comp, &op, &m_outbl);
+  assert(r == 0);
+  comp->release();
+}
+
+template<typename I>
+Context *CreateRequest<I>::handle_fetch_mirror_mode(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if ((*result < 0) && (*result != -ENOENT)) {
+    lderr(m_cct) << "failed to retrieve mirror mode: " << cpp_strerror(*result) << dendl;
+
+    m_r_saved = *result;
+    remove_object_map();
+    return nullptr;
+  }
+
+  cls::rbd::MirrorMode mirror_mode_internal = cls::rbd::MIRROR_MODE_DISABLED;
+  if (*result == 0) {
+    bufferlist::iterator it = m_outbl.begin();
+    *result = cls_client::mirror_mode_get_finish(&it, &mirror_mode_internal);
+    if (*result < 0) {
+      lderr(m_cct) << "Failed to retrieve mirror mode" << dendl;
+
+      m_r_saved = *result;
+      remove_object_map();
+      return nullptr;
+    }
+  }
+
+  // TODO: remove redundant code...
+  switch (mirror_mode_internal) {
+  case cls::rbd::MIRROR_MODE_DISABLED:
+  case cls::rbd::MIRROR_MODE_IMAGE:
+  case cls::rbd::MIRROR_MODE_POOL:
+    m_mirror_mode = static_cast<rbd_mirror_mode_t>(mirror_mode_internal);
+    break;
+  default:
+    lderr(m_cct) << "Unknown mirror mode ("
+               << static_cast<uint32_t>(mirror_mode_internal) << ")"
+               << dendl;
+    *result = -EINVAL;
+    remove_object_map();
+    return nullptr;
+  }
+
+  journal_create();
+  return nullptr;
+}
+
+template<typename I>
+void CreateRequest<I>::journal_create() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  Journal<I>::init_safetimer_instance(m_cct, &m_timer, &m_timer_lock);
+
+  // there's no need for calling ->init() as there aren't any journal
+  // objects to remove in case of failure from now on in the state
+  // machine -- header object purging is handled by ->remove()
+  m_journaler = new Journaler(m_op_work_queue, m_timer, m_timer_lock,
```

Comment:
It would probably be safer to have the journal::Create/RemoveRequest state machines instantiate and initialize the journaler object.  For example, if journal::CreateJournal fails before it inits the journal, journal::RemoveJournal will fail since the journal isn't initialized.